[RPC] Add option to make rref.get_type not block.

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #50983 [JIT/Futures] Support set_exception API on python future.
* #50979 [torch.futures] Fix doc inconsistency about callback args
* #50978 [torch.futures] Clarify callback behavior when future is completed
* **#50977 [RPC] Add option to make rref.get_type not block.**

Adds a `blocking` flag that can be set to False to make this API return a `Future` to the type. This is to make this function non-blocking, mostly for a future change that will allow `rref.rpc_async()` to be completely non-blocking (it currently calls and waits for this function that issues an RPC in-line). 

Differential Revision: [D25944582](https://our.internmc.facebook.com/intern/diff/D25944582/)

rohan-varma committed:
[RPC] Add option to make rref.get_type not block.

Per title. This allows it not to block

Differential Revision: [D25944582](https://our.internmc.facebook.com/intern/diff/D25944582/)

[ghstack-poisoned]

mrshenli commented on torch/distributed/rpc/api.py:
```diff
@@ -361,17 +362,31 @@ def _to_worker_info(to):
+        # or owner.
+        future = Future()
+        future.set_result(rref_type)
+        return future
```
if this returns a future, do we need to annotate this function with an `@rpc.functions.async_execution` decorator?

mrshenli replied:
I recall Future is not picklable, and we didn't plan to make Future picklable, as communicating a future to a remote process does not seem reasonable? E.g., if the future is not completed yet, do we need to update both local and remote futures when the local one is marked as completed?

rohan-varma replied:
This won't return a future when called remotely, as the only remote call is from the below `_rref_typeof_on_user`, and we call that without blocking (just need to get the type in that case).

I didn't decorate it with rpc.functions.async_execution for that reason, because its always ran synchronously when called over RPC. So I guess this means picklability of Future isn't a concern?

I think this could potentially cause issues later if this function is ran over RPC with `blocking=False`, but since its private I'm assuming use cases will be like the ones below.

mrshenli commented on torch/csrc/distributed/rpc/py_rref.cpp:
```diff
@@ -251,17 +251,17 @@ py::object PyRRef::createRRefProxy(
+                      : typeFuncs.onUser_(*this, timeout, blocking);
   }
-
+  // Returns py::object that can Python type or future.
```
can -> can be

mrshenli reviewed (commented):
Does it work if we always let the functions like `getRRefType` and `_rref_typeof_on_owner` return a future, and chain subsequent logic as a callback? It the Future is already completed when returning, subsequent logic will be run inline, which is the same as the current behavior. Otherwise, it's async.

rohan-varma committed:
Update on "[RPC] Add option to make rref.get_type not block."


Adds a `blocking` flag that can be set to False to make this API return a `Future` to the type. This is to make this function non-blocking, mostly for a future change that will allow `rref.rpc_async()` to be completely non-blocking (it currently calls and waits for this function that issues an RPC in-line). 

Differential Revision: [D25944582](https://our.internmc.facebook.com/intern/diff/D25944582/)

[ghstack-poisoned]

rohan-varma commented:
> Does it work if we always let the functions like `getRRefType` and `_rref_typeof_on_owner` return a future, and chain subsequent logic as a callback? It the Future is already completed when returning, subsequent logic will be run inline, which is the same as the current behavior. Otherwise, it's async.

This should work, will add unittests.

rohan-varma committed:
Update on "[RPC] Add option to make rref.get_type not block."


Adds a `blocking` flag that can be set to False to make this API return a `Future` to the type. This is to make this function non-blocking, mostly for a future change that will allow `rref.rpc_async()` to be completely non-blocking (it currently calls and waits for this function that issues an RPC in-line). 

Differential Revision: [D25944582](https://our.internmc.facebook.com/intern/diff/D25944582/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[RPC] Add option to make rref.get_type not block."


Adds a `blocking` flag that can be set to False to make this API return a `Future` to the type. This is to make this function non-blocking, mostly for a future change that will allow `rref.rpc_async()` to be completely non-blocking (it currently calls and waits for this function that issues an RPC in-line). 

Differential Revision: [D25944582](https://our.internmc.facebook.com/intern/diff/D25944582/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[RPC] Add option to make rref.get_type not block."


Adds a `blocking` flag that can be set to False to make this API return a `Future` to the type. This is to make this function non-blocking, mostly for a future change that will allow `rref.rpc_async()` to be completely non-blocking (it currently calls and waits for this function that issues an RPC in-line). 

Differential Revision: [D25944582](https://our.internmc.facebook.com/intern/diff/D25944582/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[RPC] Add option to make rref.get_type not block."


Adds a `blocking` flag that can be set to False to make this API return a `Future` to the type. This is to make this function non-blocking, mostly for a future change that will allow `rref.rpc_async()` to be completely non-blocking (it currently calls and waits for this function that issues an RPC in-line). 

Differential Revision: [D25944582](https://our.internmc.facebook.com/intern/diff/D25944582/)

[ghstack-poisoned]

rohan-varma commented:
@mrshenli Do you mind taking another look at this one? Thanks!

rohan-varma committed:
Update on "[RPC] Add option to make rref.get_type not block."


Adds a `blocking` flag that can be set to False to make this API return a `Future` to the type. This is to make this function non-blocking, mostly for a future change that will allow `rref.rpc_async()` to be completely non-blocking (it currently calls and waits for this function that issues an RPC in-line). 

Differential Revision: [D25944582](https://our.internmc.facebook.com/intern/diff/D25944582/)

[ghstack-poisoned]

mrshenli reviewed (approved):
LGTM! Shall we create an issue to track followup changes?

rohan-varma committed:
Update on "[RPC] Add option to make rref.get_type not block."


Adds a `blocking` flag that can be set to False to make this API return a `Future` to the type. This is to make this function non-blocking, mostly for a future change that will allow `rref.rpc_async()` to be completely non-blocking (it currently calls and waits for this function that issues an RPC in-line). 

Differential Revision: [D25944582](https://our.internmc.facebook.com/intern/diff/D25944582/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[RPC] Add option to make rref.get_type not block."


Adds a `blocking` flag that can be set to False to make this API return a `Future` to the type. This is to make this function non-blocking, mostly for a future change that will allow `rref.rpc_async()` to be completely non-blocking (it currently calls and waits for this function that issues an RPC in-line). 

Differential Revision: [D25944582](https://our.internmc.facebook.com/intern/diff/D25944582/)

[ghstack-poisoned]
